src/source/Ice/IceAgent.c

```diff
@@ -995,16 +981,31 @@ STATUS iceAgentStateCheckConnectionTimerCallback(UINT32 timerId, UINT64 currentT
     PStunPacket pStunBindingRequest = NULL;
     UINT64 data;
     PIceCandidatePair pIceCandidatePair = NULL;
-    UINT32 i;
+    UINT32 i, localCandidateCount, remoteCandidateCount;
+    UINT64 currTime;
 
     CHK(pIceAgent != NULL, STATUS_NULL_ARG);
 
     MUTEX_LOCK(pIceAgent->lock);
     locked = TRUE;
 
-    // extend endtime if no candidate pair is available or startIceAgent has not been called, in which case we would not
-    // have the remote password.
-    if (pIceAgent->candidatePairCount == 0 || !pIceAgent->agentStarted) {
+    currTime = GETTIME();
+
+    // if ice agent started, enforce a timeout on candidate pair creation in case no remote ice candidate is received.
+    if (pIceAgent->agentStarted && pIceAgent->candidatePairCount == 0) {
+        if (IS_VALID_TIMESTAMP(pIceAgent->candidateGenerationEndTime) && currTime >= pIceAgent->candidateGenerationEndTime) {
+            CHK_STATUS(doubleListGetNodeCount(pIceAgent->localCandidates, &localCandidateCount));
+            CHK_STATUS(doubleListGetNodeCount(pIceAgent->remoteCandidates, &remoteCandidateCount));
+            DLOGE("Failed to generate any ice candidate pair before timeout. Local candidate count: %u, Remote candidate count: %u",
+                  localCandidateCount, remoteCandidateCount);
+            CHK(FALSE, STATUS_ICE_NO_AVAILABLE_ICE_CANDIDATE_PAIR);
```

Comment:
These could be combined into CHK_ERR(FALSE, "failed to generate blah", localCandidateCont, remoteCount);